drivers/retention/retention_nrf.c

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright (c) 2023, Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nordic_nrf_retention
+
+#include <string.h>
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/retention.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(retention_nordic_nrf, CONFIG_RETENTION_LOG_LEVEL);
+
+#define NORDIC_NRF_RETENTION_REGISTER_DEFAULT_VALUE 0x00
+
+struct nordic_nrf_retention_data {
+	struct k_mutex lock;
+};
+
+struct nordic_nrf_retention_config {
+	uint32_t addr;
+	uint32_t size;
+};
+
+static int32_t nordic_nrf_retention_init(const struct device *dev)
+{
+	struct nordic_nrf_retention_data *data = dev->data;
+
+	k_mutex_init(&data->lock);
+
+	return 0;
+}
+
+static int32_t nordic_nrf_retention_size(const struct device *dev)
+{
+	const struct nordic_nrf_retention_config *config = dev->config;
+
+	return config->size;
+}
+
+static int32_t nordic_nrf_retention_read(const struct device *dev, size_t offset,
+					 uint8_t *buffer, size_t size)
+{
+	const struct nordic_nrf_retention_config *config = dev->config;
+	struct nordic_nrf_retention_data *data = dev->data;
+
+	if ((offset + size) > config->size) {
```

Comment:
Probably worth using the overflow detection helpers from include/zephyr/sys/math_extras.h throughout.

Reply:
Those functions are for overflows of variable, this overflow check is for if e.g. someone tries reading 5 bytes from offset 6 in an area which is 10 bytes, so it overflows the area but not the variable type